#Перевірка статиcтичних гіпотез. Z-test.



0. Зчитайте дані з `data.csv` у змінну data, яка має тип pandas.DataFrame. Ми будемо далі працювати з цією змінною.

In [45]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from statsmodels.stats import weightstats as stests

In [7]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [10]:
data = pd.read_csv('drive/MyDrive/Data_analitic/Python/data.csv')


Запустіть код нижче. Для коректної робити всіх подальших методів дані мають бути у вигляді одновимірного `numpy.ndarray` вектора та мати розмірність (100,). Така розмірність означає, що дані - одновимірні (колонка), якщо б розмірність була (100, 1), то дані сприймаються методами як двовимірні (таблиця), хоча для нас виглядати можуть так само.

In [11]:
data = data.data.values
data.shape

(100,)

**Завдання 1**. Зчитані дані - це сума покупок в доларах наших покупців на сайті протягом липня 2025 року.

До цього місяця, ми мали продажі в середньому на чек $20.

Необхідно зʼясувати, чи вийшло у нас статистично значущо **підвищити** середній чек за липень 2025?

Аби дати відповідь - ми проведемо z-test на рівні значущості $\alpha=0.05$ двома способами. В цьому завданні ми будемо виконувати обчислення "вручну" з використанням бібліотек numpy та scipy.stats подібно до прикладу в лекції. Для цього виконайте наступні 7 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть параметри задачі у змінні Python:
    - `sample_mean` - $\bar{x}$, середнє значення в вибірці
    - `population_mean` - $\mu_0$, середнє значення в популяції (тобто те, з яким ми порівнюємо середнє вибіркове значення)
    - `population_std` - $\sigma$, вибіркове стандартне відхилення, яке ми вважаємо, що є рівним ст. відх. популяції, адже маємо достатньо велику вибірку.
    - `sample_size` - $n$, розмір вибірки
    - `alpha` - $\alpha$ рівень значущості
    
    **Увага!** Для обчислення стандартного відхилення маємо скористатись функцією `np.std(your_dataframe, ddof=1)`. Чому так - розбираємо в лекції про t-test.


2. Визначте для цієї задачі:
    - якою є гіпотеза $H_0$
    - якою є альтернативна гіпотеза $H_a$
    - з яким типом тесту ми маємо справу - лівосторонній, правосторонній чи двосторонній.

3. Обчисліть стандартну помилку SE.
4. Розрахуйте z-статистику.
5. Знайдіть критичне z-значення з допомогою бібліотеки stats.
6. Обчисліть p-value з допомогою бібліотеки stats.
7. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$. Для прийняття рішення зробіть і порівняння z-статистики з критичним значенням, і проаналізуйте p-value.

In [23]:
sample_mean = np.round(np.mean(data),2)

In [24]:
sample_mean

np.float64(19.38)

In [32]:
sample_mean = np.mean(data)
print(f'Середнє у вибірці : {sample_mean:.2f}')
population_std = np.std(data, ddof=1)
print(f'Стандартне відхилення: {population_std:.2f}')
sample_size = len(data)
print(f'Розмір вибірки: {sample_size}')

Середнє у вибірці : 19.38
Стандартне відхилення: 6.25
Розмір вибірки: 100


# Завдання № 1

In [38]:
sample_mean = 19
population_mean = 20
population_std = 6.25
sample_size = 100
alpha = 0.05

#Завдання № 2

Гіпотези

- H₀: μ = 20 (середній чек за липень 2025)
- H₁: μ > 20 (середній чек збільшився)
- Тип тесту: правосторонній (one-tailed)

#Завдання № 3

In [35]:
population_std = 6.25
sample_size = 100
population_mean = 20

standard_error = population_std / np.sqrt(sample_size)
print(f'Стандартна помилка SE: {standard_error:.3f}')


Стандартна помилка SE: 0.625


#Завдання № 4

In [40]:
z_score = (sample_mean - population_mean) / standard_error
print(f'Z-статистика: {z_score:.3f}')

Z-статистика: -1.600


#Завдання № 5

In [41]:
z_critical = stats.norm.ppf(1 - alpha)
print(f'Z-критичне: {z_critical:.3f}')

Z-критичне: 1.645


#Завдання № 6

In [42]:
p_value = 1 - stats.norm.cdf(z_score)
print(f'p_value: {p_value:.3f}')

p_value: 0.945


#Завдання № 7

In [44]:
z_score = -1.600
z_critical = 1.645
p_value = 0.945
alpha = 0.05

# Перевірка за z-статистикою
if z_score > z_critical:
    print("Відхиляємо нульову гіпотезу на основі z-статистики.")
else:
    print("Не відхиляємо нульову гіпотезу на основі z-статистики.")

# Перевірка за p-значенням
if p_value < alpha:
    print("Відхиляємо нульову гіпотезу на основі p-значення.")
else:
    print("Не відхиляємо нульову гіпотезу на основі p-значення.")


Не відхиляємо нульову гіпотезу на основі z-статистики.
Не відхиляємо нульову гіпотезу на основі p-значення.


**Завдання 2.** Виконайте обчислення z-test з використанням бібліотеки statsmodels.

Отримайте z-статистику та р-значення.

Виведіть p-значення та зробіть висновок, чи ми маємо достатньо доказів, аби стверджувати, що середній чек зріс.

Чи зійшлись значення z-статистику та р-значення в цьому завданні з попередніми обчисленнями?

Завдання оцінюється в 3 бали.

In [46]:
z_stat_sm, p_val_sm = stests.ztest(data,
                                  value=population_mean,
                                  alternative='larger')
print(f"  Z-статистика: {z_stat_sm:.3f}")
print(f"  P-value: {p_val_sm:.6f}")

  Z-статистика: -0.995
  P-value: 0.840216


У попередніх ручних обчисленнях було:
- Z-статистика: -1.600
- p-value: близько 0.945
А при автоматичному обчисленні:
- Z-статистика: -0.995
- p-value: 0.840216

- Значення не тотожні, але мають спільну тенденцію і невелику різницю. Не має достатньо доказів, щоб стверджувати, що середній чек збільшився. Тобто не відхиляємо H₀ гіпотезу.